src/main/java/com/iexec/worker/feign/LoginService.java

```diff
@@ -49,45 +51,59 @@ public String getToken() {
 
     /**
      * Log in the Scheduler.
-     * Caution: this is NOT thread-safe.
+     * <p>
+     * Thread safety is implemented with a {@link Semaphore} and a {@code try {} finally {}} block.
+     * The lock is acquired before entering the {@code try} block.
+     * The latter has been added to ensure the lock will always be released once acquired.
+     * If the lock is not acquired, a login procedure is already ongoing and the method returns immediately.
```

Comment:
We should probably add the procedure if such a case happens: the calling thread should wait some time before trying to access the newly-created token. This may be implemented through an event (`NewLoginTokenEvent`)?

Reply:
I think this is not as straight-forward as you make it seems.
The worker implementation is mostly based on scheduled calls.
We have 2 situations:
* Some calls will be executed again after a fixed delay, nothing needs to be done
* Some calls need to implement their own wait retry mechanism

I don't see how your suggested event fits in this. Can you be more precise ?

Reply:
As discussed, we will see about this one step at a time. this seems rather dependent on the worker lifecycle and the different authenticated calls made on the Scheduler.